zio-http/src/main/scala/zio/http/Request.scala

```diff
@@ -1,199 +1,55 @@
+  def addTrailingSlash: Request = self.copy(url = self.url.addTrailingSlash)
 
   /**
-   * Decodes the body as a Body
+   * Drops trailing slash from the path.
    */
-  def body: Body
-
-  def copy(
-    version: Version = self.version,
-    method: Method = self.method,
-    url: URL = self.url,
-    headers: Headers = self.headers,
-  ): Request = {
-    val m = method
-    val u = url
-    val h = headers
-    val v = version
-    new Request {
-      override final val method: Method   = m
-      override final val url: URL         = u
-      override final val headers: Headers = h
-      override final val version: Version = v
-      override final val body: Body       = self.body
+  def dropTrailingSlash: Request = self.copy(url = self.url.dropTrailingSlash)
 
-      override final val unsafe: UnsafeAPI = new UnsafeAPI {
-        override final def context(implicit unsafe: Unsafe): Ctx        = self.unsafe.context
-        override final def encode(implicit unsafe: Unsafe): HttpRequest = self.unsafe.encode
-      }
-    }
-  }
+  val path = url.path
 
-  /**
-   * Remove trailing slash from path.
-   */
-  final def dropTrailingSlash: Request = self.copy(url = self.url.dropTrailingSlash)
+  def withMethod(method: Method): Request = self.copy(method = method)
 
-  /**
-   * Gets all the headers in the Request
-   */
-  def headers: Headers
+  def withPath(path: Path): Request = self.copy(url = self.url.copy(path = path))
 
-  /**
-   * Checks is the request is a pre-flight request or not
-   */
-  def isPreflight: Boolean = method == Method.OPTIONS
-
-  /**
-   * Gets the request's method
-   */
-  def method: Method
-
-  /**
-   * Gets the request's path
-   */
-  def path: Path = url.path
-
-  /**
-   * Overwrites the method in the request
-   */
-  final def setMethod(method: Method): Request = self.copy(method = method)
+  def withUrl(url: URL): Request = self.copy(url = url)
 
   /**
-   * Overwrites the path in the request
-   */
-  final def setPath(path: Path): Request = self.copy(url = self.url.copy(path = path))
-
-  /**
-   * Overwrites the url in the request
-   */
-  final def setUrl(url: URL): Request = self.copy(url = url)
-
-  /**
-   * Gets the complete url
-   */
-  def url: URL
-
-  /**
-   * Gets the request's http protocol version
+   * Updates the headers using the provided function
    */
-  def version: Version
-
-  private[zio] trait UnsafeAPI {
-
-    /**
-     * Accesses the channel's context for more low level control
-     */
-    def context(implicit unsafe: Unsafe): Ctx
-
-    /**
-     * Gets the HttpRequest
-     */
-    def encode(implicit unsafe: Unsafe): HttpRequest
-  }
-
-  private[zio] val unsafe: UnsafeAPI
+  override def updateHeaders(update: Headers => Headers): Request = self.copy(headers = update(self.headers))
 }
 
 object Request {
 
   /**
-   * Constructor for Request
+   * Constructor for Request that provides sane defaults.
    */
-  def apply(
+  def make(
```

Comment:
It's an anti-pattern (in my view) for a straightforward case class to have a `make` constructor. Rather, the default case class constructor should always be used: `Request(...)` (ordinary Scala) instead of `Request.make(...)` (weird Scala).

Reply:
I try to avoid default arguments in the base class itself, that's what I created a `make` so as to support sane defaults.

At previous gigs, we considered default values in a case class an anti-pattern.

Reply:
I also agree, but let's do that via `Request.default.copy(...)`.

Reply:
Or maybe `Request.get` would be more appropriate. Then a user could do: `Request.get.copy(...)`.